### PR TITLE
Feature/18 participate form

### PR DESCRIPTION
**Pull request by klarmannsonia**
In this PR:
- participate button and template with additional project information
- verify participation modal
- endpoints for submit participation and verify participation
- send email with confirmation link to user
- send email with participation confirmed to project owner

image

### CODE REVIEW

**Comment by klarmannsonia** [comment]
There is currently one warning because of the modal from the participation page. I have been trying to fix it for a few hours but did not find a solution. Maybe one of you has one. Otherwise, we can remove the modal, and after submit just redirect to the project page. 

I have send you the mailjet env keys in our channel, so that you can also test the emails.

File: backend-zs/controllers/projects.controller.js [hunk 1]
```diff
@@ -34,6 +37,75 @@ exports.createProject = function (req, res) {
...
+                }]
+            })
+
+            request.then(_ => { res.json('participant submission send successfully'); })
+                .catch(error => res.status(500).json({ error: error.message }))
+        })
+        .catch(error => res.status(500).json({ error: error.message }))
+}
+
+exports.verifyParticipation = async function (req, res) {
```
Comment (ahmad-diaa) [inline backend-zs/controllers/projects.controller.js:71]: Can we use a different statusCode for the unsuccessful responses

File: backend-zs/controllers/projects.controller.js [hunk 2]
```diff
@@ -34,6 +37,75 @@ exports.createProject = function (req, res) {
...
+        const token = req.params.token;
+        const projectId = req.params.projectId;
+        const project = await Project.findOne({ _id: projectId });
+        if (!project) {
+            return res.json({ success: false, message: 'Project not found' })
+        }
+
+        let foundParticipant
+        project.participants.forEach(participant => {
+            if (participant.confirmationToken === token) {
```
Comment (ahmad-diaa) [inline backend-zs/controllers/projects.controller.js:82]: Add a check that the confirming user is the same as the participant

File: frontend-zs/src/components/projects/SingleProject/ProjectParticipation.js [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+import Axios from 'axios';
+import { Link, useHistory } from "react-router-dom";
+import { authenticationService } from '../../../services/authentication.service';
+
+export default function ProjectParticipation(props) {
+    let history = useHistory();
+
+    const [images, setImages] = useState({})
+
+    const [project, setProject] = useState({})
```
Comment (daliamaarek) [inline frontend-zs/src/components/projects/SingleProject/ProjectParticipation.js]: Is this being used?

File: frontend-zs/src/components/projects/SingleProject/VerifyParticipation.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                    setModalTitle('Fehler');
+                    setModalBody('Ein unvorhergesehener Fehler ist aufgetreten!');
+                }
+            }
+            verifyParticipation(id, token);
+        }
+        catch (e) {
+            console.log(e);
+        }
+    }, [])
```
Comment (daliamaarek) [inline frontend-zs/src/components/projects/SingleProject/VerifyParticipation.js]:
```suggestion
    }, [props.match.params])
```

File: backend-zs/controllers/projects.controller.js (hunk 1 above)
Comment (klarmannsonia) [reply backend-zs/controllers/projects.controller.js:71]: Changed to **501** for `findOneAndUpdate` and **503** for `request`.

**501** Not Implemented
The server either does not recognize the request method, or it lacks the ability to fulfil the request.

**503** Service Unavailable
The server cannot handle the request.

File: frontend-zs/src/components/projects/SingleProject/ProjectParticipation.js (hunk 3 above)
Comment (klarmannsonia) [reply frontend-zs/src/components/projects/SingleProject/ProjectParticipation.js]: No, I deleted it.

File: backend-zs/controllers/projects.controller.js (hunk 1 above)
Comment (ahmad-diaa) [reply backend-zs/controllers/projects.controller.js:71]: I was referring to the following responses:

`return res.json({ success: false, message: 'Project not found' })`  at lines 71, 82, 87, 106. Sorry if I wasn't clear about it.

File: backend-zs/controllers/projects.controller.js [hunk 5]
```diff
@@ -34,6 +37,75 @@ exports.createProject = function (req, res) {
...
+                "To": [{
+                    "Email": projectOwner.email,
+                }],
+                "Subject": `Teilnahme am Project ${project.info.title} bestätigen`,
+                "HTMLPart": `Eine Teilnahme am Projekt ${project.info.title} wurde bestätigt. <br> <br> Teilnehmer: ${foundParticipant.name} <br> Mehr Informationen: ${foundParticipant.information}<br> Kontakt: ${foundParticipant.contact}`
+            }]
+        })
+        await res.json({ success: true, message: 'Successfully confirmed participation' });
+    } catch (e) {
+        console.log(e);
```
Comment (ahmad-diaa) [inline backend-zs/controllers/projects.controller.js]:
```suggestion
```

File: backend-zs/controllers/projects.controller.js (hunk 1 above)
Comment (ahmad-diaa) [reply backend-zs/controllers/projects.controller.js:71]: @klarmannsonia did you check my last reply here. I am afraid you resolved it before I posted my reply.

**Review by daliamaarek — APPROVED** [review]

File: backend-zs/controllers/projects.controller.js [hunk 6]
```diff
@@ -40,6 +43,74 @@ exports.createProject = function (req, res) {
...
+
+        let foundParticipant
+        project.participants.forEach(participant => {
+            if (participant.confirmationToken === token) {
+                participant.isConfirmed = true;
+                foundParticipant = participant;
+            }
+        })
+        if (!foundParticipant || foundParticipant.user != req.body._id) {
+            return res.json({ success: false, message: 'Participant for project not found' })
```
Comment (ahmad-diaa) [inline backend-zs/controllers/projects.controller.js]: Those are the responses I was referring to regarding changing the status codes. along with lines 77, 93, 109

File: backend-zs/controllers/projects.controller.js [hunk 7]
```diff
@@ -40,6 +43,74 @@ exports.createProject = function (req, res) {
...
+                    "Email": "felix@zukunftschreiben.org",
+                },
+                "To": [{
+                    "Email": projectOwner.email,
+                }],
+                "Subject": `Teilnahme am Project ${project.info.title} bestätigen`,
+                "HTMLPart": `Eine Teilnahme am Projekt ${project.info.title} wurde bestätigt. <br> <br> Teilnehmer: ${foundParticipant.name} <br> Mehr Informationen: ${foundParticipant.information}<br> Kontakt: ${foundParticipant.contact}`
+            }]
+        })
+        await res.json({ success: true, message: 'Successfully confirmed participation' });
```
Comment (ahmad-diaa) [inline backend-zs/controllers/projects.controller.js]: Is the `await` needed here?

File: backend-zs/controllers/projects.controller.js (hunk 6 above)
Comment (klarmannsonia) [reply backend-zs/controllers/projects.controller.js]: Fixed

File: backend-zs/controllers/projects.controller.js (hunk 7 above)
Comment (klarmannsonia) [reply backend-zs/controllers/projects.controller.js]: Fixed

File: backend-zs/controllers/projects.controller.js [hunk 8]
```diff
@@ -74,7 +74,7 @@ exports.verifyParticipation = async function (req, res) {
         const projectId = req.params.projectId;
         const project = await Project.findOne({ _id: projectId });
         if (!project) {
-            return res.json({ success: false, message: 'Project not found' })
+            return res.status(501).json({ success: false, message: 'Project not found' })
```
Comment (ahmad-diaa) [inline backend-zs/controllers/projects.controller.js]:
```suggestion
            return res.status(404).json({ success: false, message: 'Project not found' })
```

`404` goes better with entity not found

File: backend-zs/controllers/projects.controller.js [hunk 9]
```diff
@@ -85,12 +85,12 @@ exports.verifyParticipation = async function (req, res) {
             }
         })
         if (!foundParticipant || foundParticipant.user != req.body._id) {
-            return res.json({ success: false, message: 'Participant for project not found' })
+            return res.status(502).json({ success: false, message: 'Participant for project not found' })
```
Comment (ahmad-diaa) [inline backend-zs/controllers/projects.controller.js]:
```suggestion
            return res.status(404).json({ success: false, message: 'Participant for project not found' })
```

File: backend-zs/controllers/projects.controller.js [hunk 10]
```diff
@@ -85,12 +85,12 @@ exports.verifyParticipation = async function (req, res) {
...
         }
 
         const projectOwner = await User.findById(project.projectOwner)
         if (!projectOwner) {
-            return res.json({ success: false, message: 'Project owner not found' })
+            return res.status(503).json({ success: false, message: 'Project owner not found' })
```
Comment (ahmad-diaa) [inline backend-zs/controllers/projects.controller.js]:
```suggestion
            return res.status(404).json({ success: false, message: 'Project owner not found' })
```

File: backend-zs/controllers/projects.controller.js [hunk 11]
```diff
@@ -40,6 +43,74 @@ exports.createProject = function (req, res) {
...
+                "To": [{
+                    "Email": projectOwner.email,
+                }],
+                "Subject": `Teilnahme am Project ${project.info.title} bestätigen`,
+                "HTMLPart": `Eine Teilnahme am Projekt ${project.info.title} wurde bestätigt. <br> <br> Teilnehmer: ${foundParticipant.name} <br> Mehr Informationen: ${foundParticipant.information}<br> Kontakt: ${foundParticipant.contact}`
+            }]
+        })
+        res.status(200).json({ success: true, message: 'Successfully confirmed participation' });
+    } catch (e) {
+        await res.status(500).json({ success: false, message: e })
```
Comment (ahmad-diaa) [inline backend-zs/controllers/projects.controller.js]:
```suggestion
       res.status(500).json({ success: false, message: e })
```

**Review by ahmad-diaa — APPROVED** [review]

**Pull request merged**